Source/JavaScriptCore/wasm/WasmSectionParser.cpp

```diff
@@ -491,7 +503,7 @@ auto SectionParser::parseElement() -> PartialResult
         case 0x05: {
             Type refType;
             WASM_PARSER_FAIL_IF(!parseRefType(m_info, refType), "can't parse reftype in elem section");
-            WASM_PARSER_FAIL_IF(!refType.isFuncref(), "reftype in element section should be funcref");
+            WASM_PARSER_FAIL_IF(!isFuncref(refType) && refType.isNullable(), "reftype in element section should be funcref");
```

Comment:
I am confused about the logic here, but also the message is wrong?

Reply:
Yeah, this is unfortunately a bit confusing.

When the typed function references proposal is fully fleshed out, it will support both tables and element sections with concrete function types (as described in https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#summary). But right now, it's restricted to `funcref`.

The confusing thing is that the representation of `funcref` changes a bit between Wasm with reference types and Wasm with typed function references. In the former, it was always a type with `TypeKind::Funcref`. In the latter, it is a `TypeKind::Ref` with a heap type of `func` and the other form becomes a shorthand for this.

The old code `!refType.isFuncref()` is only checking for the shorthand form. This causes some of the tests in this patch to fail, because some of the binary modules use the new form. The new code is doing the equivalent check but allowing for the non-shorthand `Ref` type form. So there is no change in functionality here; the code is just accounting for a change in representation caused by typed function references.